app/models/grda_warehouse/tasks/scrub_pii/scrub_client_pii_task.rb

```diff
@@ -0,0 +1,115 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module GrdaWarehouse::Tasks::ScrubPii
+  # * Selectively scrub personally identifiable information (PII) for HMIS Clients
+  # * Delete client-related custom hmis data
+  # * Delete versions
+  class ScrubClientPiiTask
+    def self.perform(...)
+      new.perform(...)
+    end
+
+    def perform(client_ids: nil, data_source_ids: nil, custom_scrubber: nil, progress: false)
+      with_lock do
+        @scrubber = Pii::Scrubber::ScrubModelPii.new(custom_scrubber: custom_scrubber, progress: progress)
+
+        client_scope = GrdaWarehouse::Hud::Client.with_deleted
+        client_scope = client_scope.where(id: client_ids) if client_ids
+        client_scope = client_scope.where(data_source_id: data_source_ids) if data_source_ids
+
+        @scrubber.perform(client_scope)
+        client_scope.find_in_batches do |clients|
+          process_client_batch(clients)
+        end
+      end
+    end
+
+    protected
+
+    def process_client_batch(clients)
+      client_ids = clients.map(&:id)
+
+      # delete directly-related client records
+      delete_custom_data_elements(Hmis::Hud::Client, client_ids)
+      delete_versions([GrdaWarehouse::Hud::Client, Hmis::Hud::Client], client_ids)
+
+      clients.group_by(&:data_source_id).map do |data_source_id, ds_clients|
+        personal_ids = ds_clients.map(&:PersonalID)
+        delete_hmis_client_records(personal_ids, data_source_id)
+        delete_enrollment_related_records(personal_ids, data_source_id)
+      end
+    end
+
+    def delete_custom_data_elements(model, ids)
```

Comment:
This could have a clearer name to indicate that it's not wholesale deleting ALL the custom data elements on the model. `delete_custom_data_elements_with_pii`?

Reply:
Good suggestion, I'll rename it. Perhaps we can replace this in the future with a more formal formal pii tracking field, such as  `Hmis::Hud::CustomDataElement.where(pii_level: ..2)`